src/scheduler.cpp

```diff
@@ -23,113 +23,86 @@
 
 void Scheduler::threadMain()
 {
-	std::unique_lock<std::mutex> eventLockUnique(eventLock, std::defer_lock);
-	while (getState() != THREAD_STATE_TERMINATED) {
-		std::cv_status ret = std::cv_status::no_timeout;
-
-		eventLockUnique.lock();
-		if (eventList.empty()) {
-			eventSignal.wait(eventLockUnique);
-		} else {
-			ret = eventSignal.wait_until(eventLockUnique, eventList.top()->getCycle());
-		}
-
-		// the mutex is locked again now...
-		if (ret == std::cv_status::timeout && !eventList.empty()) {
-			// ok we had a timeout, so there has to be an event we have to execute...
-			SchedulerTask* task = eventList.top();
-			eventList.pop();
-
-			// check if the event was stopped
-			auto it = eventIds.find(task->getEventId());
-			if (it == eventIds.end()) {
-				eventLockUnique.unlock();
-				delete task;
-				continue;
-			}
-			eventIds.erase(it);
-			eventLockUnique.unlock();
-
-			task->setDontExpire();
-			g_dispatcher.addTask(task, true);
-		} else {
-			eventLockUnique.unlock();
-		}
-	}
+	io_service.run();
 }
 
 uint32_t Scheduler::addEvent(SchedulerTask* task)
 {
-	eventLock.lock();
-
-	if (getState() != THREAD_STATE_RUNNING) {
-		eventLock.unlock();
-		delete task;
-		return 0;
-	}
-
 	// check if the event has a valid id
 	if (task->getEventId() == 0) {
+		uint32_t tmpEventId = ++lastEventId;
 		// if not generate one
-		if (++lastEventId == 0) {
-			lastEventId = 1;
+		if (tmpEventId == 0) {
+			tmpEventId = ++lastEventId;
 		}
-
-		task->setEventId(lastEventId);
+		task->setEventId(tmpEventId);
 	}
 
 	// insert the event id in the list of active events
-	uint32_t eventId = task->getEventId();
-	eventIds.insert(eventId);
-
-	// add the event to the queue
-	eventList.push(task);
+	io_service.post([this, task]() {
+		boost::asio::deadline_timer* timer;
+		if (timerCacheVec.empty()) {
+			timer = new boost::asio::deadline_timer(io_service);
+		} else {
+			timer = timerCacheVec.back();
+			timerCacheVec.pop_back();
+		}
+		eventIdTimerMap[task->getEventId()] = timer;
```

Comment:
I get when you are reusing the same timer objects here. Is the construction of a timer that heavy that you can't create a new one every time, and delete it when done?